packages/react/src/ReactElementValidator.js

```diff
@@ -259,6 +261,105 @@ function validateFragmentProps(fragment) {
   setCurrentlyValidatingElement(null);
 }
 
+export function jsxWithValidation(
+  isStaticChildren,
+  type,
+  props,
+  key,
+  source,
+  self,
+) {
+  const validType = isValidElementType(type);
+
+  // We warn in this case but don't throw. We expect the element creation to
+  // succeed and there will likely be errors in render.
+  if (!validType) {
+    let info = '';
+    if (
+      type === undefined ||
+      (typeof type === 'object' &&
+        type !== null &&
+        Object.keys(type).length === 0)
+    ) {
+      info +=
+        ' You likely forgot to export your component from the file ' +
+        "it's defined in, or you might have mixed up default and named imports.";
+    }
+
+    const sourceInfo = getSourceInfoErrorAddendum(source);
+    if (sourceInfo) {
+      info += sourceInfo;
+    } else {
+      info += getDeclarationErrorAddendum();
+    }
+
+    let typeString;
+    if (type === null) {
+      typeString = 'null';
+    } else if (Array.isArray(type)) {
+      typeString = 'array';
+    } else if (type !== undefined && type.$$typeof === REACT_ELEMENT_TYPE) {
+      typeString = `<${getComponentName(type.type) || 'Unknown'} />`;
+      info =
+        ' Did you accidentally export a JSX literal instead of a component?';
+    } else {
+      typeString = typeof type;
+    }
+
+    warning(
+      false,
+      'React.jsx: type is invalid -- expected a string (for ' +
+        'built-in components) or a class/function (for composite ' +
+        'components) but got: %s.%s',
+      typeString,
+      info,
+    );
+  }
+
+  const element = jsx.apply(this, Array.prototype.slice.call(arguments, 1));
```

Comment:
If we do the thing where jsxDEV is its own special thing, then we don't need this to be called with arity. Just call it jsx(type, props, key).

Reply:
I believe ElementValidator only runs in dev, so this would actually call jsxDEV with source and self. Check the newest code - not sure it's the best way, but I think it has the ideal api, where jsxDEV is opt in for source/self but jsx in dev will still give you all of the warnings.